### PR TITLE
[feature] add KV cache manager for llama & bloom inference

### PR DESCRIPTION
**Pull request by yuanheng-zhao**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

This PR adds KV cache manager used for llama and bloom model inference (adapted from lightly memory manager), and adds an inference state info data class that will be used during inference.

together work with @tiandiao123 @CjhHa1 @yuehuayingxueluo 

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/inference/kvcache_manager.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    def free(self, free_index):
+        self.available_size += free_index.shape[0]
+        self.mem_state[free_index] = 1
+
+    @torch.no_grad()
+    def free_all(self):
+        self.available_size = len(self.mem_state)
+        self.mem_state[:] = 1
+        self.past_key_values_length = 0
+        self.logger.info("freed all space of memory manager")
```
Comment (CjhHa1) [inline colossalai/inference/kvcache_manager.py:92]: maybe add some basic tests for the `infer_state`

File: colossalai/inference/inference_state.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+# might want to consider combine with InferenceConfig in colossalai/ppinference/inference_config.py later
+from dataclasses import dataclass
+from typing import Any
+
+import torch
+
+from colossalai.inference.kvcache_manager import MemoryManager
+
+
+@dataclass
```
Comment (CjhHa1) [inline colossalai/inference/inference_state.py:10]: why not use infer_info :)

File: colossalai/inference/kvcache_manager.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+        self._init_mem_states(size, device)
+        self._init_kv_buffers(size, device, dtype, head_num, head_dim, layer_num)
+
+    def _init_mem_states(self, size, dev):
+        self.mem_state = torch.ones((size,), dtype=torch.bool, device=dev)
+        self._mem_cum_sum = torch.empty((size,), dtype=torch.int32, device=dev)
+        self.indexes = torch.arange(0, size, dtype=torch.long, device=dev)
+
+    def _init_kv_buffers(self, size, dev, dtype, head_num, head_dim, layer_num):
```
[the remark this answers is not in the document]
Comment (yuanheng-zhao) [reply colossalai/inference/kvcache_manager.py:40]: Guess that's a trivial point. There exist many vars representing device in Colossalai naming as `src_dev`, `cur_dev`, `dev`, `init_dev`, etc.

File: colossalai/inference/kvcache_manager.py (hunk 1 above)
Comment (yuanheng-zhao) [reply colossalai/inference/kvcache_manager.py:92]: will add some tests for `InferenceState` and kvcache manager

File: colossalai/inference/inference_state.py (hunk 2 above)
Comment (yuanheng-zhao) [reply colossalai/inference/inference_state.py:10]: To refactor the class and add other attributes we would like to use as well as some utility functions

File: colossalai/inference/kvcache_manager.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yuanheng-zhao) [reply colossalai/inference/kvcache_manager.py]: modified

File: colossalai/inference/kvcache_manager.py (hunk 1 above)
Comment (yuanheng-zhao) [reply colossalai/inference/kvcache_manager.py:116]: Tests for kvcache manager added. Batch inference state will be tested with llama/bloom inference flow.

**Review by tiandiao123 — APPROVED** [review]

**Comment by tiandiao123** [comment]
LGTM！

File: tests/test_infer/test_kvcache_manager.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+INPUT_LEN = 16
+OUTPUT_LEN = 8
+LAYER_NUM = 4
+HEAD_NUM = 32
+HEAD_DIM = 128
+
+
+def create_cache_manager(rank, world_size, port, batch_size, input_len, output_len, layer_num, head_num, head_dim):
+    os.environ['RANK'] = str(rank)
+    os.environ['LOCAL_RANK'] = str(rank)
```
Comment (CjhHa1) [inline tests/test_infer/test_kvcache_manager.py:20]: Why the `rank` and `local rank` are same?

File: tests/test_infer/test_kvcache_manager.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    value_buffers = kvcache_manager.value_buffer
+    assert len(key_buffers) == len(value_buffers) == layer_num
+    assert key_buffers[0].shape == value_buffers[0].shape
+    # required size exceeds the maximum allocated size
+    invalid_locs = kvcache_manager.alloc_contiguous(size + 1)
+    assert invalid_locs is None
+    # for prefill stage, allocation via alloc and alloc_contiguous should be the same
+    total_token_prefill = batch_size * input_len
+    prefill_locs = kvcache_manager.alloc(total_token_prefill)
+    kvcache_manager.free_all()
```
Comment (CjhHa1) [inline tests/test_infer/test_kvcache_manager.py:38]: `free` func has not been tested ,and may be we can check the `available size` after free

File: tests/test_infer/test_kvcache_manager.py (hunk 4 above)
Comment (yuanheng-zhao) [reply tests/test_infer/test_kvcache_manager.py:20]: dumbly set the same for testing

**Pull request merged**